### DISCUSSION TITLE
Question about connection and how to get token

### DISCUSSION THREAD

**Discussion by szopenkrk** [TonyM1958/FoxESS-Cloud#12] [General]
Hi  @TonyM1958  its gr8 option that You are building that solution I'm fighting now with integration with FoxEss Cloud all the time i get data: { errno: 41809, result: null }

I have something like that 

```
function md5c(text = "", _type = "lower") {
    const hash = crypto.createHash("md5").update(text, "utf-8").digest("hex");
    console.log("Hash", hash);
    return _type === "lower" ? hash.toLowerCase() : hash.toUpperCase();
  }
```

 ```
 function generateSignature(apiKey, domain) {
    let timestamp = Date.now();
    let signatureRaw = `${domain}\r\n${apiKey}\r\n${timestamp}`;
    let signature = md5c(signatureRaw);

    return {
      token: apiKey,
      lang: "en",
      timestamp: timestamp,
      signature: signature,
      "User-Agent": "Chrome/117.0.0.0 Safari/537.36",
    };

```
```
axios
    .get(`https://www.foxesscloud.com/c/v0/user/access`, { headers })
    .then((response) => {
      console.log("response Fox", response);
      res.json(response.data);
    })
    .catch((error) => {
      console.error("There was an error!", error);
      res.status(500).send("Error fetching Login from FoxEss");
    });
```
`

**Comment by TonyM1958** [comment]
I think you need to review the Open API documentation here: https://www.foxesscloud.com/public/i18n/en/OpenApiDocument.html

So far as I can see from your code:
+ you are using an Open API key with a Fox Cloud endpoint instead of an Open API end point. Open API end points start with /op/
+ timestamp sent to the API needs to be milliseconds. I use round(time.time() * 1000) to get this

There are examples of the code to generate the correct header in my Open API module.

Error codes are listed here: https://github.com/TonyM1958/FoxESS-Cloud/wiki/Error-Codes
41809: Illegal Token.

**Reply by szopenkrk** [reply]
I start dig in in python code and from examples of open api foxess its working i need to check whole code of my node.js thanks

**Reply by szopenkrk** [reply]
For future questions in js we need to do 
``${path}\\r\\n${token}\\r\\n${timestamp}`` 

and its working

**Comment by delanthear** [comment]
Just following up with my struggles with this on PHP.  I'm getting a similar error, but in PHP.  

```
$token = "<snip>";   //api token
$serial = "<snip>"; // serial of inverter

$foxessapi = "https://www.foxesscloud.com";
$path = "/op/v0/device/generation";
$params = "sn=$serial";

$timestamp = intval(round(microtime(true) *1000));
$signature = $path . "\\r\\n" . $token . "\\r\\n" . $timestamp;
$md5_sig = md5($signature);

$request_headers = [
	'Content-Type: application/json',
	'token:' . $token,
	'signature:' . $md5_sig,
	'timestamp:' . $timestamp,
	'lang:en'
];
```

Curl returns this:   {"errno":40256,"msg":"Parameter could not be parsed correctly illegal signature"}

**Reply by szopenkrk** [reply]
Use that its my js code 

```
let domain = "https://www.foxesscloud.com";
const key = "your key";

function getSignature(token, path, lang = "en") {
  const timestamp = Date.now();
  const signatureString = `${path}\\r\\n${token}\\r\\n${timestamp}`;
  const signature = md5c(signatureString);
  console.log(path);
  return {
    token: token,
    lang: lang,
    timestamp: timestamp,
    signature: signature,
    "User-Agent":
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36",
  };
}

function md5c(text) {
  const hash = crypto.createHash("md5").update(text, "utf8").digest("hex");
  return hash.toLowerCase();
}

app.get("/api/foxess/plant/list", async (req, res) => {
  const path = "/op/v0/plant/list";
  const headers = getSignature(key, path);
  try {
    const response = await axios.post(
      `${domain}${path}`,
      { currentPage: 1, pageSize: 10 },
      { headers }
    );
    res.json(response.data);
  } catch (error) {
    console.error(error);
    res.status(500).send("Server error");
  }
});

```

**Reply by szopenkrk** [reply]
Yes there is an issue during setup of Your signature I had the same and spend almost 2 days on that in my js code it works perfectly

**Reply by delanthear** [reply]
As far as I can tell, I'm doing the same as you, just in PHP.  Either I'm missing something obvious or its how I'm doing the timestamp or the md5?

**Reply by TonyM1958** [reply]
@delanthear Simplest thing to do is to start with something that is working and see where your code is going wrong. Open Google Colab and put this code in a new notebook:

```
!pip install foxesscloud --root-user-action=ignore --quiet --upgrade
import foxesscloud.openapi as f
f.api_key = "<your key>"
f.debug_setting=3
f.get_generation()
```

Add your Open API key inside the quotes where it says "\<your key\>".

Assuming you just have 1 inverter, the first time you run this, it will go through a number of API calls to get your device details and will get the generation data. If you run it again, it will just get the generation data. If you have multiple inverters, you can add a line with a full or part inverter serial number like this after the API key:

```
f.device_sn="<your sn>"
```

With the debug setting, the code will print out the API calls, the parameters it is working with and the headers when it runs. Copy the parameters and use these for testing your PHP code to make sure it generates the signature correctly. Once you replicate the signature / headers, your API call should work.

**Reply by delanthear** [reply]
What I don't get is that if I change the timestamp somehow (like * 100 instead of 1000) the timestamp parameter gets rejected.  This suggests to me that the timestamp is correctly formatted as it's being accepted on it's own.  

That leaves the path, which is clearly correct, and the token, which I know is correct, the md5, which seems fairly simple (php returns it as a 32character hex value, which I think is the same as yours), or the \r\n joining characters.  I've tried them as \r\n or escaped as \\r\\n which outputs as text when printed, rather than as the carriage returns and newlines.  

There isn't else for me to look at here!  The answer must be something stupid :D

**Reply by delanthear** [reply]
Thanks for that Tony.  Putting the details into Colab and running it and it works.  If I, within a minute, take the token, timestamp and signature that were working in Colab and put them in the script, I get the same error.     This is skipping the MD5, the \r\n etc, basically hardcoding it.  Perhaps something wrong in the curl request with the headers?!

**Comment by delanthear** [comment]
Hey Tony, is there are way to dump the http call out in the Colab example so I can compare when what's being logged in my curl log?  Just incase there is a difference in the http call.

**Reply by delanthear** [reply]
Ok hold on that question.  It's just literally started working with no changes!   It wasn't working just now, so  I added curl_getinfo call to see if there was anything suspicious and then it worked.  Reverted that change back to how it was before I added the getinfo and it's still working....

Perhaps something changed on Fox's site? *boggle*

**Reply by TonyM1958** [reply]
> Hey Tony, is there are way to dump the http call out in the Colab example so I can compare when what's being logged in my curl log? Just incase there is a difference in the http call.

If you add this code to a cell in your notebook before you run the cell I gave you, it will turn on http logging and debug output from the http client:

```
import logging
import http.client as http_client
http_client.HTTPConnection.debuglevel = 1
logging.basicConfig()
logging.getLogger().setLevel(logging.DEBUG)
requests_log = logging.getLogger("requests.packages.urllib3")
requests_log.setLevel(logging.DEBUG)
requests_log.propagate = True
```

**Comment by Meintre** [comment]
Hello TomM1958,
Unfortunately I don't speak English (Google Translate).
I have the following problem: Everything works as it should under Google Colab.
"!pip install foxesscloud --root-user-action=ignore --quiet --upgrade
import foxesscloud.openapi as f
f.api_key = "75932c48-9b4c-43be-a1ff-xxxxxxxxx"
f.device_sn = "60AT5020238xxxx"
f.time_zone = "Europe/London"
f.get_generation()" -> {'month': 31.89999999999636, 'today': 0.5, 'cumulative': 10161.9}

But if I do the same thing on a Mac in the terminal:
1. "pip3 install foxesscloud --root-user-action=ignore --quiet --upgrade"
2. in the file it says (python3 Foxess.py):
"import foxesscloud.openapi as f
f.api_key = "75932c48-9b4c-43be-a1ff-xxxxxxxxxx"
f.device_sn = "60AT5020238xxxx"
f.time_zone = "Europe/London"
f.get_generation()" -> FoxESS Cloud Open API version 2.7.3
** device model not recognized for deviceType: T5-G3

Can you help me?

**Reply by TonyM1958** [reply]
I can have a look but I may not be able to reproduce the problem. Please can you add this line after the import:

```
f.debug_setting=2
```

and run get_device() instead of get_generation() and post the results?

**Comment by Meintre** [comment]
FoxESS-Cloud Open API version 2.7.3
getting messages
getting variables
getting device
getting flag
** get_flag(), no result data, errno = 41935: The device cannot be operated offline
getting generation
** device model not recognised for deviceType: T5-G3
imacmeini@iMacMeini desktop %

**Reply by TonyM1958** [reply]
Thanks - that's the error I expect to see, not sure why you are not seeing the same error in Google Colab. It's because the code doesn't currently correctly identify the attributes for a T series inverter so I will need to make a change for that.

**Reply by TonyM1958** [reply]
I've pushed v2.7.4 out but it will take about 15 minutes before this gets out to all the distribution servers. Give this a try and see if it works for you.

**Comment by Meintre** [comment]
this is the result -> 

FoxESS-Cloud Open API version 2.7.4
getting messages
getting variables
getting device
getting flag
Traceback (most recent call last):
  File "/Users/imacmeini/Desktop/foxess.py", line 11, in <module>
    f.get_device()
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/foxesscloud/openapi.py", line 479, in get_device
    get_flag()
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/foxesscloud/openapi.py", line 1009, in get_flag
    if errno == 40256:
       ^^^^^
NameError: name 'errno' is not defined. Did you forget to import 'errno'?

**Reply by TonyM1958** [reply]
Thanks - that was a silly coding error that slipped through my testing. Have another go with v2.7.5.

**Comment by Meintre** [comment]
this is the result at Google Colab with this code 

"!pip3 install foxesscloud --root-user-action=ignore --quiet --upgrade
import foxesscloud.openapi as f
f.debug_setting=2
f.api_key = "75932c48-9b4c-43be-a1ff-xxxxxxxx"
f.device_sn = "60AT5020238xxxxxx"
f.time_zone = "Europe/London"

f.get_device()" ->

{'deviceType': 'T5-G3',
 'masterVersion': '1.24',
 'afciVersion': '0.19',
 'hasPV': True,
 'deviceSN': '60AT5020238xxxx',
 'slaveVersion': '1.00',
 'hasBattery': False,
 'function': {'scheduler': False},
 'hardwareVersion': '--',
 'managerVersion': '1.12',
 'stationName': 'Trenkmann',
 'moduleSN': '609W3DGF233xxxx',
 'productType': 'T3',
 'stationID': 'd5d81f07-d08d-47a6-bcc0-xxxxxxx',
 'status': 3,
 'generationToday': 0.5,
 'generationTotal': 10161.9,
 'eps': False}

**Reply by TonyM1958** [reply]
You should get the same error here - it looks like Google Colab may have exception handling disabled.

**Comment by Meintre** [comment]
Good morning, you probably never sleep!
Thanks for the quick reply.
This is the result on the Mac ->

FoxESS-Cloud Open API version 2.7.5
getting messages
getting variables
getting device
getting flag
getting generation
getting generation
imacmeini@iMacMeini desktop % 

and this at Google colab ->

{'deviceType': 'T5-G3',
 'masterVersion': '1.24',
 'afciVersion': '0.19',
 'hasPV': True,
 'deviceSN': '60AT5020238xxxx',
 'slaveVersion': '1.00',
 'hasBattery': False,
 'function': {'scheduler': False},
 'hardwareVersion': '--',
 'managerVersion': '1.12',
 'stationName': 'Trenkmann',
 'moduleSN': '609W3DGF233xxxx',
 'productType': 'T3',
 'stationID': 'd5d81f07-d08d-47a6-bcc0-xxxxxxx',
 'status': 3,
 'generationToday': 0.0,
 'generationTotal': 10161.9,
 'eps': False,
 'model': 'T3',
 'phase': 3,
 'power': 5.0,
 'max_charge_current': None}

**Reply by TonyM1958** [reply]
Good morning, looks like that should be working OK for you now?

It's difficult to decode the power rating and phase info until I know the exact format of the device type code that Fox returns as I don't have access to lots of inverter models for testing. The original error was the code displaying the device type so it could be added.

**Comment by Meintre** [comment]
Thanks again, let's wait until it's online.
I would send you the access data for my inverter
if you need it!

**Comment by Meintre** [comment]
Hello TomM1958,
Unfortunately I still can't get any data from the MacBook via the terminal.
It works via Google Colab.
Can you check again?

Mac File -> 
"import foxesscloud.openapi as f
f.api_key = "75932c48-9b4c-43be-a1ff-af87ebxxxx"
f.device_sn = "60AT5020238Cxxx"
f.time_zone = "Europe/London"

f.get_real("PV1Power") "

Output: 

imacmeini@iMacMeini desktop % python3 foxess.py
FoxESS-Cloud Open API version 2.7.5
imacmeini@iMacMeini desktop % 

Google Colab File ->

"!pip install foxesscloud --root-user-action=ignore --quiet --upgrade
import foxesscloud.openapi as f
f.api_key = "75932c48-9b4c-43be-a1ff-af87eb6xxxx"
f.device_sn = "60AT5020238Cxxx"
f.time_zone = "Europe/London"

f.get_real("pv1Power") "

Output: 

getting real-time data
[{'unit': 'kW',
  'name': 'PV1Power',
  'variable': 'pv1Power',
  'value': 0.09809999999999999}]

**Reply by TonyM1958** [reply]
Thanks. I wonder if the output is being redirected or you need to explicitly print the result, where Jupyter Lab implicitly sends the result of the final statement to the console?

If you are on FB and in the Fox ESS Owners group, find me in the list of moderators and message me if you are happy to privately share your API key so I can try running the script in a shell.

**Comment by Meintre** [comment]
I logged into Facebook but unfortunately couldn't find the Fox ESS Owners group.
My name is Meinhard Trenkmann

**Reply by TonyM1958** [reply]
I've accepted your request to join the group so you should be to see me if you go to the group and click on Members and look at the Admins and Moderators.

**Comment by Meintre** [comment]
I found you and sent you the data

**Reply by TonyM1958** [reply]
Thanks. I've accessed your inverter and run this script:

```
import openapi as f
import private as my

f.api_key = my.fox_api_key3

result = f.get_generation()
print(f"generation today = {result['today']:.1f} kWh")
print(f"generation month = {result['month']:.1f} kWh")
print(f"generation total = {result['cumulative']:.1f} kWh")

result = f.get_real(['pv1Power', 'pvPower'])
print(result)
```

The output from this looks like this:

![image](https://github.com/user-attachments/assets/fe997bff-79e1-4d28-a409-8d2bf39a8305)

The import statements are different for me as I'm working in developer mode with direct access to the code. As you can see, it prints the generation data but doesn't return the current power data as it's offline when dark.

Please can you try running the main part of this with the explicit print statements on your Mac and see if you get the same results?

Here's what I get when I run this against my inverter:

![image](https://github.com/user-attachments/assets/7626ea5b-d57b-4d2e-ae75-b612d8f22a58)

If this still doesn't work for you, I'll try setting up a Mac Book and have another look at what it might be doing.

**Comment by Meintre** [comment]
This is my version (simply brilliant). Thank you very much. This is what I needed. I'll have the real data from PV1 tomorrow and I've got a Christmas present!

-> FoxESS-Cloud Open API version 2.7.5
generation today = 0.6 kWh
generation month = 32.5 kWh
generation total = 10162.5 kWh
** get_real(): device 60AT5020238C045 is not on-line, status = off-line (3)
None
imacmeini@iMacMeini desktop %

**Reply by TonyM1958** [reply]
No problem. There is a lot more, potentially, that you can do. For example, this script runs when the inverter is offline and shows you the PV Power:

```
# list of dates to use. today=0 will get yesterdays complete data, today=1 will get todays partial data, or enter a specific date in 'YYYY-MM-DD' format
d = f.date_list(span='2days', today=1)

# get the data from Fox Cloud
result = f.get_history('day', d=d, v=['pv1Power'], plot=1)

# print the results
for r in result:
  print(f"{r['date']} / {r['name'].replace('Power', '')}: Peak output was {r['max']:.2f} kW at {r['max_time']}. Total energy generated was {r['kwh']:.1f} kWh")
```

Produces this output for your inverter:

![image](https://github.com/user-attachments/assets/1e4433e0-43dc-46ed-b6db-02061ffb87b9)

The plots may not work on the Mac Book but there is a way to save the images to files if needed.

**Comment by Meintre** [comment]
That promises a lot of interesting tasks.
But I'm going away this weekend, so I won't be able to do it until next week. I'm also using Iobroker for a lot of diagrams. But I'll try it out next week and then report back!
Have a nice weekend and thank you again.
